source/extensions/clusters/aggregate/cluster.h

```diff
@@ -0,0 +1,124 @@
+          Runtime::RandomGenerator& random,
+          Server::Configuration::TransportSocketFactoryContext& factory_context,
+          Stats::ScopePtr&& stats_scope, ThreadLocal::SlotAllocator& tls, bool added_via_api);
+
+  // Upstream::Cluster
+  Upstream::Cluster::InitializePhase initializePhase() const override {
+    return Upstream::Cluster::InitializePhase::Secondary;
+  }
+
+  // Upstream::ClusterUpdateCallbacks
+  void onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) override;
+  void onClusterRemoval(const std::string& cluster_name) override;
+
+  void refresh() {
+    refresh([](const std::string&) { return false; });
+  }
+
+  std::vector<std::string> clusters_;
+  Upstream::ClusterUpdateCallbacksHandlePtr handle_;
+  Upstream::ClusterManager& cluster_manager_;
+  Runtime::Loader& runtime_;
+  Runtime::RandomGenerator& random_;
+  ThreadLocal::SlotPtr tls_;
+
+private:
+  // Upstream::ClusterImplBase
+  void startPreInit() override;
+
+  void refresh(std::function<bool(const std::string&)>&& skip_predicate);
+  std::pair<Upstream::PrioritySetImpl,
+            std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>
+  linearizePrioritySet(std::function<bool(const std::string&)> skip_predicate);
+};
+
+using PriorityContext = std::pair<Upstream::PrioritySetImpl,
+                                  std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>;
+class LoadBalancerImpl : public Upstream::LoadBalancerBase {
+public:
+  LoadBalancerImpl(Cluster& cluster, PriorityContext&& priority_context)
+      : Upstream::LoadBalancerBase(priority_context.first, cluster.info()->stats(),
+                                   cluster.runtime_, cluster.random_, cluster.info()->lbConfig()),
+        priority_to_cluster_(std::move(priority_context.second)) {}
+
+  // Upstream::LoadBalancer
+  Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext* context) override;
+
+  // Upstream::LoadBalancerBase
+  Upstream::HostConstSharedPtr chooseHostOnce(Upstream::LoadBalancerContext*) override {
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE
+  }
+
+private:
+  std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>> priority_to_cluster_;
+};
+using LoadBalancerImplSharedPtr = std::shared_ptr<LoadBalancerImpl>;
+
+class AggregateClusterLoadBalancer : public Upstream::LoadBalancer {
+public:
+  // Upstream::LoadBalancer
+  Upstream::HostConstSharedPtr chooseHost(Upstream::LoadBalancerContext* context) override;
+
+  void refresh(LoadBalancerImplSharedPtr load_balancer) { load_balancer_ = load_balancer; }
+
+private:
+  LoadBalancerImplSharedPtr load_balancer_;
+};
+
+struct AggregateLoadBalancerFactory : public Upstream::LoadBalancerFactory {
+  // Upstream::LoadBalancerFactory
+  Upstream::LoadBalancerPtr create() override {
+    return std::make_unique<AggregateClusterLoadBalancer>();
+  }
+};
+
+struct AggregateThreadAwareLoadBalancer : public Upstream::ThreadAwareLoadBalancer {
```

Comment:
there are a lot of -LoadBalancer/Factory classes in play here, i think this could warrant some comments explaining how they all work together